go/consensus/tendermint/api/state.go

```diff
@@ -0,0 +1,245 @@
+		state:         ms.tree,
+		appState:      ms,
+		blockHeight:   ms.cfg.BlockHeight,
+		blockCtx:      ms.blockCtx,
+		logger:        logging.GetLogger("consensus/tendermint/abci").With("mode", mode),
+	}
+	c.Context = context.WithValue(context.Background(), contextKey{}, c)
+
+	return c
+}
+
+// NewMockApplicationState creates a new mock application state for testing.
+func NewMockApplicationState(cfg MockApplicationStateConfig) ApplicationState {
+	tree := mkvs.New(nil, nil)
+
+	blockCtx := NewBlockContext()
+	if cfg.MaxBlockGas > 0 {
+		blockCtx.Set(GasAccountantKey{}, NewGasAccountant(cfg.MaxBlockGas))
+	} else {
+		blockCtx.Set(GasAccountantKey{}, NewNopGasAccountant())
+	}
+
+	return &mockApplicationState{
+		cfg:      cfg,
+		blockCtx: blockCtx,
+		tree:     tree,
+	}
+}
+
+// ImmutableState is an immutable state wrapper.
+type ImmutableState struct {
+	Tree mkvs.KeyValueTree
+}
+
+// CheckContextMode checks if the passed context is an ABCI context and is using one of the
+// explicitly allowed modes.
+func (s *ImmutableState) CheckContextMode(ctx context.Context, allowedModes []ContextMode) error {
+	abciCtx := FromCtx(ctx)
+	if abciCtx == nil {
+		return fmt.Errorf("abci: method must only be called from ABCI context")
+	}
+
+	for _, m := range allowedModes {
+		if abciCtx.Mode() == m {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("abci: method cannot be called from the specified ABCI context mode (%s)", abciCtx.Mode())
+}
+
+// Close releases the resources associated with the immutable state wrapper.
+//
+// After calling this method, the immutable state wrapper should not be used anymore.
+func (s *ImmutableState) Close() {
+	if tree, ok := s.Tree.(mkvs.ClosableTree); ok {
+		tree.Close()
+	}
+}
+
+// NewImmutableState creates a new immutable state wrapper.
+func NewImmutableState(ctx context.Context, state ApplicationState, version int64) (*ImmutableState, error) {
+	if state == nil {
+		return nil, ErrNoState
+	}
+
+	// Check if this request was made from an ABCI application context.
+	if abciCtx := FromCtx(ctx); abciCtx != nil {
+		// Override used state with the one from the current context in the following cases:
+		//
+		// - If this request was made from InitChain, no blocks and states have been submitted yet.
+		// - If this request was made from an ABCI app and is for the current (future) height.
+		//
+		if abciCtx.IsInitChain() || version == abciCtx.BlockHeight()+1 {
+			return &ImmutableState{Tree: abciCtx.State()}, nil
+		}
+	}
+
+	// Handle a regular (external) query where we need to create a new tree.
+	if state.BlockHeight() == 0 {
+		return nil, consensus.ErrNoCommittedBlocks
+	}
+	if version <= 0 || version > state.BlockHeight() {
+		version = state.BlockHeight()
+	}
+
+	ndb := state.Storage().NodeDB()
+	roots, err := ndb.GetRootsForRound(ctx, uint64(version))
+	if err != nil {
+		return nil, err
+	}
+	if len(roots) != 1 {
+		return nil, fmt.Errorf("state: corrupted state (%d): %+v", version, roots)
```

Comment:
I think I already explained this -- so our MKVS supports multiple tree roots per version (height in this case). This is because for runtimes we store both IO roots and state roots and there can be multiple of them until the version is finalized (because until discrepancy detection happens, you don't know which root is correct). This is also the reason why `Commit` and `Finalize` are split.

Of course in the case where MKVS is used for consensus layer state there is always only one root per version so this looks strange. I was thinking we could do some optimizations for this case and also enforce it at the database level by having a `MaxRootsPerVersion` config option.

Reply:
ok. thanks for the explanation